pubsub/natspubsub/nats_test.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2019 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package natspubsub
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/driver"
+	"gocloud.dev/pubsub/drivertest"
+
+	"github.com/nats-io/gnatsd/server"
+	gnatsd "github.com/nats-io/gnatsd/test"
+	"github.com/nats-io/go-nats"
+)
+
+const (
+	TEST_PORT  = 11222
+	BENCH_PORT = 9222
+)
+
+type harness struct {
+	s  *server.Server
+	nc *nats.Conn
+}
+
+func newHarness(ctx context.Context, t *testing.T) (drivertest.Harness, error) {
+	opts := gnatsd.DefaultTestOptions
+	opts.Port = TEST_PORT
+	s := gnatsd.RunServer(&opts)
+	nc, err := nats.Connect(fmt.Sprintf("nats://127.0.0.1:%d", TEST_PORT))
+	if err != nil {
+		return nil, err
+	}
+	return &harness{s, nc}, nil
+}
+
+func (h *harness) CreateTopic(ctx context.Context, testName string) (dt driver.Topic, cleanup func(), err error) {
+	cleanup = func() {}
+	dt = openTopic(h.nc, testName)
+	return dt, cleanup, nil
+}
+
+func (h *harness) MakeNonexistentTopic(ctx context.Context) (driver.Topic, error) {
+	// A nil *topic behaves like a nonexistent topic.
+	return (*topic)(nil), nil
+}
+
+func (h *harness) CreateSubscription(ctx context.Context, dt driver.Topic, testName string) (ds driver.Subscription, cleanup func(), err error) {
+	ds = openSubscription(h.nc, testName)
+	// FIXME(dlc) - Check for error?
+	cleanup = func() {
+		var sub *nats.Subscription
+		if ds.As(&sub) {
+			sub.Unsubscribe()
+		}
+	}
+	return ds, cleanup, nil
+}
+
+func (h *harness) MakeNonexistentSubscription(ctx context.Context) (driver.Subscription, error) {
+	return (*subscription)(nil), nil
+}
+
+func (h *harness) Close() {
+	if h == nil {
```

Comment:
Why are there all these nil checks? Seems fine to just let them panic if they are nil since it's a legitimate bug.

Reply:
Can remove.